[ADD][9.0] l10n_fi / Finnish localization module

mlaitinen opened the pull request:
Description of the issue/feature this PR addresses:

Added Finnish localization module for 9.0
- Contains `res.partner` fields for Finnish business ID (validated) and EDI code
- Can generate a payment reference (domestic or international) for sales invoices
## 

I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

mlaitinen commented:
Your contribution guidelines state:

> Important: Please do NOT create pull requests with the same patch on multiple target branches.
> Fixes made in LTS versions (currently 7.0, 8.0, 9.0) will be forward ported to later versions periodically.

Please let me know if you'll accept my backport to 8.0 (doesn't exist yet).

Yenthe666 commented:
@mart-e this one is for you!
@mlaitinen Odoo can backport changes or fixes to older versions if they make sense and don't break the rules (such as adding new fields).

mart-e commented:
Sure adding a localization is fine in stable version as explained in the version table but the sentence you quote implied that you should make the request to the lowest branch first so we avoid backporting when not needed.
Anyway it's a new module so I can merge it in 8.0 directly instead (no need to create a new PR).

Also it seems you have not [signed the CLA](https://www.odoo.com/sign-cla).
Could you add it so we can integrate your patch?

mart-e commented on addons/l10n_fi/__openerp__.py:
```diff
@@ -0,0 +1,29 @@
+# noinspection PyStatementEffect
+{
+    "name": "Finnish localization",
+    "version": "9.0.0.1.0",
```
prefer the version number formating `1.0`, otherwise it's a bit painful to port

mart-e commented:
Well actually the accounting has changed a lot between 8 and 9 so you will need to make two pull requests if the data is different. What I explained above was correct if it's exactly the same code.

mart-e commented on addons/l10n_fi/__openerp__.py:
```diff
@@ -0,0 +1,29 @@
+    "category": "Localization",
+    "description": """
+        General modifications useful to any Finnish company.
+        See README.md for details.
```
Why not put everything in the description? The README file is not available to user that install a module, only to system administrators.

mart-e commented:
@qdp-odoo this module introduces new fields and code (which we try to avoid in l10n modules) so your call !

mlaitinen commented:
Thanks for your comments and reacting so quickly!


CLA: In fact I have signed the CLA, both [Individual](https://github.com/odoo/odoo/blob/9.0/doc/cla/individual/mlaitinen.md) and [Corporate](https://github.com/odoo/odoo/blob/9.0/doc/cla/corporate/avoinsystems.md). My GitHub account is registered with my gmail address while my code was committed with my company address. Should I close this PR and create a new one with correct author data?

Version: I will change the version from 9.0.0.1.0 to 1.0

New fields and code: Please please please accept the fields I have created. They are essential for 90 % of Finnish companies and I wouldn't want to create an OCA module just for that.

README: The mention about `README.md` was accidentially left there. I removed `README.md` right before creating this PR and copied the information to `__openerp__.py` and `static/description/index.html`. I'll remove the mention to `README.md`

mart-e commented:
> CLA: In fact I have signed the CLA, both Individual and Corporate. My GitHub account is registered with my gmail address while my code was committed with my company address. Should I close this PR and create a new one with correct author data?

Right, I remembered correctly you already signed it :) The issue is that your corporate CLA is signed with `miku.laitinen@` email address while you commit with an `miku@` address.
So you can either:
- modify the commits to use the email from the corporate signature
- add the second address to your corporate signature

> New fields and code: Please please please accept the fields I have created. They are essential for 90 % of Finnish companies and I wouldn't want to create an OCA module just for that.

The issue when adding new field/code is that it will be applied to all records, even the one that do not use finish accounting (e.g. multicompany). It's usually a bit painful to maintain.
For instance I see you add a `business_id` field. Couldn't you use the `base_vat` module for that?

mlaitinen commented:
@mart-e I fixed the author email & version and removed the reference to `README.md`. All checks seem to have passed.

It seems that `base_vat` doesn't have an appropriate field for the business ID. All companies in Finland have a business ID but only companies that are found in the VAT register have a VAT ID. 

AFAIK there are a lot of implementations of the Finnish business ID and we've been talking with other Finnish Odoo vendors that we should get rid of the non-standard implementations and create a single official implementation that we can depend on. 

Likewise, the payment reference field is also used in 99 % of the B2B _and_ B2C non-cash invoices. I know that standard and validated payment references aren't a thing in continental Europe, but in Finland pretty much all payments are made with a standard payment reference with a check digit, as specified by the banks.

treviser commented:
@mlaitinen Sorry to barge in, but isn't the Business ID the same as the company registry, as the trade register's registration number was replaced by the Business ID in 2001? Here in Switzerland for example I use the company registry for the trade register's business identification and the VAT ID for the VAT identification.

mlaitinen commented:
@treviser: Sorry, I didn't quite get you. What are you saying?

treviser commented:
@mart-e mentioned that it is difficult, when you want to add new fields, such as the field business_id. According to my understanding, the finnish Business ID should be entered in the existing field "company_registry" in the object "res.company". This is the place where you should enter the official trade registry's company registration ID.

mlaitinen commented:
Yes @treviser, that's exactly where we put the Business ID of our own company. In this module I have added the exactly same field (with validation) for `res.partner` because many companies need to know the Business ID of their clients.

treviser commented:
Ah, stupid me, that's something different. Sorry for the bothering.

mlaitinen commented:
Is there any chance this PR ever gets merged? Would it better suit for `master`? There's a true need for this module and now I'm considering opening a PR to OCA's `l10n-finland` to speed things up. It's a shame because I'd really like to see this in Odoo SaaS by default.

mart-e commented:
Hi!

Sorry for the delay. Let's take a decision on this one, shall we.
I checked it again, and actually, one issue is that it is not really a chart of account. There is no accounts, journals, taxes,... defined.
If you check with your module installed (or install account with Finland as the country), you will see that Odoo still requires a chart of account.

So while the addition of new fields is debatable but not inconceivable for a new module, we need to have at least the minimum accounts.

mlaitinen commented:
Changed my mind about this.
